design/open/iceberg-branching.md

```diff
@@ -0,0 +1,80 @@
+# Iceberg support: branching & isolation
+
+## Problem Description
+
+Iceberg stores metadata files that represent a snapshot of a given Iceberg table.
+Those metadata files save the locataion of the data hard-coded.
+This is problematic with lakeFS, since we cannot work on the same files with different branches.
+
+For example when working on branch `main`, the metadata files will be written with the property `"location" : "s3a://example-repo/main/db/table1"`.
+This means the data is located in `"s3a://example-repo/main/db/table1"`.
+However, when branching out from main to `feature-branch`, the metadata files still point to the main branch.
+With this we lose the capability of isolation.
+
+
+## Goals
+
+Enable isolation with iceberg table. Create branch from existing one, and writing the data and metadata to the relevant branch.
+At first stage: support iceberg without being catalog agnostic - meaning work with only lakeFS catalog and not be compatible with other catalogs.
+At second stage: being catalog agnostic. Users will be able to configure both lakeFS catalog and other catalog together.
+- The catalogs that will be supported are Glue, Snoflake and Tabular.
+
+## Non Goals
+
+- Support merge operation.
+- Enable lakeFS operations using a dedicated SQL syntax.
+
+
+## Proposed Design
+
+Implement lakeFS catalog:
+
+The catalog stores current metadata pointer for Iceberg tables.
+With lakeFS, this pointer needs to be versioned as well.
+In order to prevent from adding versioning capabilities to existing catalogs, we'll extend a catalog that can utilize lakeFS to keep the pointer.
+Therefor, we'll extend hadoop catalog to work with lakeFS
+- Every operation of the catalog happens with the scope of a reference in lakeFS.
+- Write location on metadata files to be relative: the repo and brach will not be written.
+- The catalog will know to use the metadata files from the relevant references
+  
+Implement lakeFSIO:
+FileIO is the primary interface between the core Iceberg library and underlying storage. In practice, it is just two operations: open for read and open for write.
+We'll Extend hadoop fileIO to work with lakeFS, and to be compatible with lakeFS catalog. 
+
```

Comment:
I know very little about Iceberg and FileIOs, only what I remember from about a year ago, so I am probably asking silly questions.  But we might want to answer (some of) them in this design doc.

1. What is the behaviour of the lakeFSIO?  Does it affect manifests, or something more?  Does it change the _path_ to the manifests?

   To pick an example at random of something I don't understand: The [manifests spec](https://iceberg.apache.org/spec/#manifests) says `data_file.file_path` is the "full URI for the file with FS scheme".  Is lakeFSIO responsible for returning it in this manner?  If so, what gets stored in the manifest object itself and how does lakeFSIO process it?  Does lakeFSIO understand Avro?
1. How do we support concurrency?  This is the point of using Iceberg, "_safely... at the same time_" [appears](https://iceberg.apache.org/) in the 2-line description at the top.  The spec goes into [atomic rename](https://iceberg.apache.org/spec/#file-system-tables).  Do we need to do that?  lakeFS _does_ support atomic rename by branching out, copying and deleting, and merging back.  Does lakeFSIO need to do that, or will we be fine without it?

   Or, Iceberg seems to want to do this to implement some [isolation between reads and writes](https://iceberg.apache.org/spec/#optimistic-concurrency).  Is the plan to use lakeFS directly for this concurrency?
1. I had a quick look and it seems that Iceberg may allow importing external files: for instance this might be an [`add_files` operation](https://github.com/apache/iceberg/issues/2068).  How does this work?  What is the intended behaviour?  If it is not supported, can we block it if we only implement a FileIO?

Reply:
1. the fileIO (with sync to the catalog) is responsible to write and read the files, it will write the relative locations in the metadata files, and will have the scope of lakeFS repo and ref. Beside that it should work the same.

2. At first stage lakeFS won't support atomicity. We believe it's Ok since we are always at scope of a brach.
We'll require a single-writer per branch. It's a limitation but we think it's reasonable for now.

3. I think that for now it's out of scope. I added it to the open questions to keep that in mind.

Reply:
Please explain (1) in the design.  I am still not sure how we propose to translate to relative paths:
* Does Iceberg _already_ use relative paths
* Or, does our FileIO perform this translation...
  - when writing?
  - when reading?

About (2) thanks for adding text!  In order of preference, we need to:
* Enforce single-writer (e.g., all writers but the first fail)
* Report multiple writers (e.g., all writers or all writers but the first report a multiple writers error, and the state is indeterminate)
* "Just" document it.
But we need to do at least the last in the list.

About (3), again: can we block it, or if not can we report it, or do we "just" document it?